### PR TITLE
new(userspace/engine): generalize indexable ruleset

### PR DESCRIPTION
**Pull request by mstemm**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

> /kind release

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

> /area tests

> /area proposals

> /area CI

**What this PR does / why we need it**:
Now that custom rules loading implementations (and related, custom rulesets) can be swapped into falco in a customizable way, there is some functionality in evttype_index_ruleset that could be used by other rulesets, specifically the part that segregates filters by ruleset and enables/disables filters based on name substring + tags. This PR makes a base class for that. See the commit comments for more details.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
NO

```release-note
NONE
```

### CODE REVIEW

File: userspace/engine/evttype_index_ruleset.cpp [hunk 1]
```diff
@@ -230,180 +66,48 @@ void evttype_index_ruleset::on_loading_complete()
...
-		m_rulesets.emplace_back(std::make_shared<ruleset_filters>());
-	}
-
-	for(const auto &wrap : m_filters)
-	{
-		bool matches;
-		std::string::size_type pos;
-
-		switch(match)
+		std::shared_ptr<wrapper> derived = std::dynamic_pointer_cast<wrapper>(wrap);
```
Comment (jasondellaluce) [inline userspace/engine/evttype_index_ruleset.cpp:73]: A dynamic cast for every rules index bucket, for every event, seems like a bit overkill for performance to me. I like this modularization effort, but can we choose a more performant route? For example, maybe the wrapper type can be a template parameter?

Comment (mstemm) [reply userspace/engine/evttype_index_ruleset.cpp]: I can static_cast it instead, the wrapper will always be the derived type anyway. I switched it to static_cast.

**Review by jasondellaluce — APPROVED** [review]
/approve

**Review by FedeDP — APPROVED** [review]
/approve

**Pull request merged**